### PR TITLE
Transport NIC config method with addresses instead of device

### PR DESCRIPTION
**Pull request by manadart**
When detecting link-layer devices and addresses (on-machine and via the instance-poller), Juju has until now transported the address configuration method with the **device**, but saved it against the **address**.

In this patch, we:
- Add `ConfigType` to the `Address` wire type.
- Ensure all providers report detected addresses with a populated `ConfigType`.
- Prefer the address member when transforming detected link-layer info to state args.

Note that the provider changes are non functional because at the time of writing we only ever decorate existing devices and address with provider IDs - we do not add them. This is expected to change in the future.

This is the last member from the **device** (prior members included CIDR) that required us to duplicate it per detected **address**. That means we can now follow this patch with a change to machine device detection that no longer does this duplication. We can simplify it and populate the slice of addresses, leaving the old cardinality mismatch behind us.

## QA steps

- Bootstrap to MAAS.
- Add a machine
- Add a LXD container to this machine, with positive space constraints.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Transport ConfigType with Addresses](https://trello.com/c/hcrPTfGe/3576-transport-configtype-with-addresses)

File: provider/ec2/environ.go [hunk 1]
```diff
@@ -1379,17 +1378,18 @@ func mapNetworkInterface(iface amzec2.NetworkInterface, subnet amzec2.Subnet) ne
...
 		if privAddr.Association.PublicIP != "" {
-			ni.ShadowAddresses = append(
-				ni.ShadowAddresses,
-				network.NewProviderAddress(
-					privAddr.Association.PublicIP, network.WithScope(network.ScopePublic)),
-			)
+			ni.ShadowAddresses = append(ni.ShadowAddresses, network.NewProviderAddress(
+				privAddr.Association.PublicIP,
+				network.WithScope(network.ScopePublic),
+				network.WithConfigType(network.ConfigDHCP),
```
Comment (achilleasa) [inline provider/ec2/environ.go:1391]: Probably doesn't change anything but why do we assume that ec2 assigns VIPs via DHCP?

**Review by achilleasa — APPROVED** [review]

File: provider/ec2/environ.go (hunk 1 above)
Comment (manadart) [reply provider/ec2/environ.go:1391]: I don't know, and it's probably incorrect anyway. If it was correct, your VM could change public IP during downtime, which seems like undesirable behaviour.

Comment (manadart) [reply provider/ec2/environ.go:1391]: As soon as I can manage, I will change the instance-poller API facade to process these addresses. I'll look at this properly at that time.

**Comment by manadart** [comment]
$$merge$$

File: provider/ec2/environ.go (hunk 1 above)
Comment (achilleasa) [reply provider/ec2/environ.go:1391]: That seems to be the case on ec2 at least. See https://github.com/juju/juju/pull/10902 and associated bug.

Comment (manadart) [reply provider/ec2/environ.go:1391]: Yes, now that I think on it, they would rather charge for assignable floating IPs than give you the "feature" for free.

**Pull request merged**